src/Shared/AWS/AWSSemanticConventions.cs

```diff
@@ -0,0 +1,329 @@
+        where T : IList<KeyValuePair<string, object>> => Add(attributes, x => x.AttributeServiceVersion, value, addIfEmpty);
+    #endregion
+
+    #region Url
+    /// <inheritdoc cref="AWSSemanticConventionsBase.AttributeUrlPath"/>
+    public static T AddAttributeUrlPath<T>(this T attributes, object? value, bool addIfEmpty = false)
+        where T : IList<KeyValuePair<string, object>> => Add(attributes, x => x.AttributeUrlPath, value, addIfEmpty);
+    /// <inheritdoc cref="AWSSemanticConventionsBase.AttributeServiceName"/>
+    public static T AddAttributeUrlQuery<T>(this T attributes, object? value, bool addIfEmpty = false)
+        where T : IList<KeyValuePair<string, object>> => Add(attributes, x => x.AttributeUrlQuery, value, addIfEmpty);
+    #endregion
+
+    private static T Add<T>(this T attributes, Func<AWSSemanticConventionsBase, string> attributeNameFunc, Func<AWSSemanticConventionsBase, string> valueFunc)
+        where T : IList<KeyValuePair<string, object>> => Add(attributes, attributeNameFunc, valueFunc(GetSemanticConventionVersion()));
+
+    private static T Add<T>(this T attributes, Func<AWSSemanticConventionsBase, string> attributeNameFunc, object? value, bool addIfEmpty = false)
+        where T : IList<KeyValuePair<string, object>>
+    {
+        var semanticConventionVersionImpl = GetSemanticConventionVersion();
+
+        var attributeName = attributeNameFunc(semanticConventionVersionImpl);
+
+        // if attributeName is empty or there is no value, exit
+        if (string.IsNullOrEmpty(attributeName) ||
+           (string.IsNullOrEmpty(value?.ToString()) && !addIfEmpty))
+        {
+            return attributes;
+        }
+
+        attributes.Add(new(attributeName, value ?? string.Empty));
+
+        return attributes;
+    }
+
+    private static Activity? SetTag(this Activity? activity, Func<AWSSemanticConventionsBase, string> attributeNameFunc, Func<AWSSemanticConventionsBase, object?> valueFunc) =>
+        SetTag(activity, attributeNameFunc, valueFunc(GetSemanticConventionVersion()));
+
+    private static Activity? SetTag(this Activity? activity, Func<AWSSemanticConventionsBase, string> attributeNameFunc, object? value)
+    {
+        var semanticConventionVersionImpl = GetSemanticConventionVersion();
+
+        var attributeName = attributeNameFunc(semanticConventionVersionImpl);
+
+        // if attributeName is empty, exit
+        if (string.IsNullOrEmpty(attributeName))
+        {
+            return activity;
+        }
+
+        activity?.SetTag(attributeName, value);
+
+        return activity;
+    }
+
+    private static IDictionary<string, string> AddDic(IDictionary<string, string> dict, Func<AWSSemanticConventionsBase, string> attributeNameFunc, string value)
+    {
+        var semanticConventionVersionImpl = GetSemanticConventionVersion();
+
+        var attributeName = attributeNameFunc(semanticConventionVersionImpl);
+
+        if (!string.IsNullOrEmpty(attributeName))
```

Comment:
Empty strings may be valid and meaningful depending on the concrete attribute.

Reply:
:)